### ISSUE TITLE
Cannot instantiate a Java Class or call vararg methods

### ISSUE THREAD

**Issue by C0D3-M4513R** [KubeJS-Mods/KubeJS#583]
### Minecraft Version

1.18.2

### KubeJS Version

1802.5.5-build.567

### Rhino Version

1802.2.1-build.255

### Architectury Version

4.10.88

### Forge/Fabric Version

Forge 40.2.0

### Describe your issue

I would expect 
```js
new java('net.minecraft.network.chat.ClickEvent')(java('net.minecraft.network.chat.ClickEvent$Action').OPEN_URL, url)
```
 to return a valid ClickEvent back. Instead I get hit with `Error occurred while handling event 'command.registry': TypeError: dev.latvian.mods.rhino.NativeJavaObject@1fbc3b1f is not a function, it is object.`

As for just calling Java Reflevction methods like so
```js
java('net.minecraft.network.chat.ClickEvent').getContructor(java('net.minecraft.network.chat.ClickEvent$Action'), java('java.lang.String')).newInstance(java('net.minecraft.network.chat.ClickEvent$Action').OPEN_URL, "url")
```
that gives a `Wrapped java.lang.IllegalArgumentException: argument type mismatch` error.

**Comment by ChiefArug** [comment]
I think for the first one the () for the constructor are trying to execute before the `new` keyword. Save the result of `java()` to a variable before trying to do anything with it. Standarrd KJS practice is to name the variable `$ClassName`.

For the second, its not working because you are passing in a `ClickEvent$Action` and a `NativeString`. Not a `ClickEvent$Action` and a `String`. 
Normally Rhino would typewrap the `NativeString` to a `String`, however it does not here because `newInstance` takes a vararg of `Object`s, so it does not see a need to.

**Comment by ChiefArug** [comment]
Also, please use the #support channel for stuff like this, like you were told in the discord. This almost certainly isn't a bug.

**Comment by ChiefArug** [comment]
Support ticket means a post in the #support channel, not an issue on Github. Issues on Github are for bugs and feature requests, of which this is neither.

**Comment by ChiefArug** [comment]
> So how is this not a bug?

Because Rhino will and does convert between NativeString and String whenever it sees a need to. The only time it doesn't is when the code could take a NativeString already, ie it takes an Object. (note that type erasure does apply here as well). 
If the method says it takes an Object there is **no way** to know if it should actually be taking something else. Rhino can only see it takes an Object.
You can force Rhino to coerce something to a String by using the String constructor that takes a String. Rhino will see that it takes a String and convert your NativeString to an actual String before passing it
```js
const $String = java('java.lang.String')
let foo_js = "foo"
let foo_java = new $String(foo_js)
```
\
I am not sure why that is still failing with that updated example. Someone will probably have to do some testing.
If it does turn out to be a bug, it will need fixing in Rhino, which is no longer getting updated because Lat is busy working on its successor, Ichor. So there is not much you can do either way unfortunately. Use `ComponentWrapper#clickEventOf` for this usecase. If there is another use case that you need reflection for ask in the Discord #support channel and we can help you figure out another way around it.

**Comment by MichaelHillcox** [comment]
What am I doing wrong here?

https://c.mikey.pro/scr/yEWWawlK.png

```
const clickEvent = java('net.minecraft.network.chat.ClickEvent');
const action = java("net.minecraft.network.chat.ClickEvent$Action")
const e = new clickEvent(action.RUN_COMMAND, "/list");

console.log(e.getAction())
console.log(e.getValue())

const PermissionTypes = java('net.minecraftforge.server.permission.nodes.PermissionTypes');
const PermissionNode = java('net.minecraftforge.server.permission.nodes.PermissionNode');

const modId = 'id';
const deletePermission = 'delete';
const node = new PermissionNode(
		modId,
		deletePermission,
		PermissionTypes.BOOLEAN,
		()=>false,
		[]
);

console.log(node.getNodeName())
console.log(node.getReadableName())
console.log(node.getDefaultResolver())
```

I can run both of your example scripts absolutely fine in 1.18.2? cu

**Comment by MichaelHillcox** [comment]
Tested in dev using the latest dev branch for 1.18.2.

This is what I used to test in a production env

![](https://c.mikey.pro/scr/94o0wTYv.png)

**Comment by squoshi** [comment]
Ideally, as Chief said, you can use the support forums on the Discord. The GitHub is to report issues with the mod itself and give feature requests, and again, as already stated, this is not a bug or a feature request.

**Comment by MichaelHillcox** [comment]
I'm testing your code now, but if this doesn't work, my assumption is it's an issue with how global is handled more than it is an issue with the construction of Java classes.

**Comment by MichaelHillcox** [comment]
Right, I can confirm, there is definitely something wrong with assigning a Java class directly to the global object. I'd assume it's an issue with the Java sided data type that backs this data store, but without looking I can't confirm. 

Here is a solid test case 

```
global.sharedClasses = {
  PermissionTypes: java(
    "net.minecraftforge.server.permission.nodes.PermissionTypes"
  ),
  PermissionNode: java(
    "net.minecraftforge.server.permission.nodes.PermissionNode"
  ),
};

global.PermissionTypes = java(
  "net.minecraftforge.server.permission.nodes.PermissionTypes"
);

(global.PermissionNode = java(
  "net.minecraftforge.server.permission.nodes.PermissionNode"
)),
  console.log(global.sharedClasses);

(() => {
  const modId = "id";
  const deletePermission = "delete";
  const node = new global.sharedClasses.PermissionNode(
    modId,
    deletePermission,
    global.sharedClasses.PermissionTypes.BOOLEAN,
    () => false,
    []
  );

  console.log(node.getNodeName());
  console.log(node.getReadableName());
  console.log(node.getDefaultResolver());
})();

console.log("Second");

(() => {
  const modId = "id";
  const deletePermission = "delete";
  const node = new global.PermissionNode(
    modId,
    deletePermission,
    global.PermissionTypes.BOOLEAN,
    () => false,
    []
  );

  console.log(node.getNodeName());
  console.log(node.getReadableName());
  console.log(node.getDefaultResolver());
})();
```

Although this is clearly an issue, I'd recommend against polluting the global space with classes regardless. My recommendation would be to do something like I have done by containing the shared classes inside a nested object.

```
global.sharedClasses = {
  PermissionTypes: java(
    "net.minecraftforge.server.permission.nodes.PermissionTypes"
  ),
  PermissionNode: java(
    "net.minecraftforge.server.permission.nodes.PermissionNode"
  ),
};
```

**Comment by MichaelHillcox** [comment]
```
  const node = new global.sharedClasses.PermissionNode(
    "id",
    "delete",
    global.sharedClasses.PermissionTypes.BOOLEAN,
    () => false,
    []
  );

  console.log(node.getNodeName());
  console.log(node.getReadableName());
  console.log(node.getDefaultResolver());
```

This code here works fine for me

**Comment by MichaelHillcox** [comment]
I'll investigate and talk to Lat, I'm pretty sure it's gunna be an issue with the backing data store, likely quite far down the stack so it'll be a pain to fix

**Comment by MaxNeedsSnacks** [comment]
Well this has been... a big thread I never noticed due to the slew of other issues; my apologies. I'll be closing this for now since Rhino has changed a fair bit in the time since this issue was made, but, if it is still an issue on newer versions (which it may be!), feel free to open an issue on the Rhino repo instead, I'm just cleaning out KubeJS issues right now

**Comment by C0D3-M4513R** [comment]
I won't be checking this on newer versions, because the underlying project, which I wrote the code for, got dissolved in the meantime.
Additionally the original kubejs scripts also got abandoned almost entirely due to this issue iirc (ref. https://github.com/KubeJS-Mods/KubeJS/issues/583#issuecomment-1430164496), as it was judged to be easier to work around the issue by reimplementing all of the logic that had been written up until that point in java.
